### PR TITLE
Support wgkex loadbalancing/client steering mechanism

### LINKED ISSUE #142: wgkex client: Add feature to get gateway from wgkex
This will help to spread all nodes over all possible gateways as wgkex can then send the gateway with the least nodes to the node requesting a gateway.

This needs code on the wgkex side as well:
https://github.com/freifunkMUC/wgkex/issues/39

### PR DESCRIPTION
**Pull request by DasSkelett**
See https://github.com/freifunkMUC/wgkex/pull/87, this is the first draft to implement the client side in checkuplink.

This PR is a placeholder for now, I guess it shouldn't be merged to `main` in the end, which seems to be kept in sync with upstream.

## Changes

- Add functionality to get gateway peer details from the wgkex broker instead of reading from site.conf
- Using this, the wgkex broker can loadbalance the clients between available gateways (https://github.com/freifunkMUC/wgkex/pull/87)
- This behaviour can be turned on by setting `loadbalancing = '1'` in site.conf
  In this case, the `peers` option is no longer needed to be set
- **The supposed value of the `broker` option has changed, it should now only be the base path (without `http://` scheme prefix and without `/` suffix) instead of the full path with `/api/v1/wg/key/exchange`**
- Added dependency on `lua-jsonc`

Will reopen in upstream repo after review.

Closes https://github.com/freifunkMUC/site-ffm/issues/142

### CODE REVIEW

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 1]
```diff
@@ -18,24 +18,42 @@ clean_port() {
...
+
+combine_ip_port() {
+    local ip="$1"
+    local port="$2"
+
+    if echo "$ip" | grep -q ":"; then
+        echo "[$ip]:$port"
+    else
+        echo "$ip:$port"
+    fi
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:39]: You probably want to use regex matching instead and maybe simplify the if/else

```suggestion
    if [[ ${ip} == *":"* ]]; then
            ip="[${ip}]"
    fi
    echo "$ip:$port"
```

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:1]: You want to run shellcheck over this script. Preferably shellcheck master, if you can build it yourself and set the shebang to `#!/bin/busybox sh`.

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 2]
```diff
@@ -18,24 +18,42 @@ clean_port() {
...
-		echo $ipv4$(echo $peer_endpoint | egrep -oe :[0-9]+$)
+		local ipv4="$(gluon-wan nslookup $peer_host | grep 'Address:\? [0-9]' | grep -oE "\b([0-9]{1,3}\.){3}[0-9]{1,3}\b")"
+		echo $ipv4
 	fi
-
 }
 
 # Do we already have a private-key? If not generate one
-temp=$(uci get wireguard.mesh_vpn.privatekey)
+uci get wireguard.mesh_vpn.privatekey >/dev/null
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:56]: This should not be needed with the latest version of checkuplink (November 2023).

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 3]
```diff
@@ -65,33 +83,78 @@ if [ "$(uci get wireguard.mesh_vpn.enabled)" == "true" ] || [ "$(uci get wiregua
...
 		wget -q https://[::1]
 		if [ $? -eq 1 ]; then
 			PROTO=http
 		else
 			PROTO=https
 		fi
-		gluon-wan wget -q  -O- --post-data='{"domain": "'"$SEGMENT"'","public_key": "'"$PUBLICKEY"'"}' $PROTO://$(uci get wireguard.mesh_vpn.broker)
-		
+
+		if [ "$(uci get wireguard.mesh_vpn.loadbalancing)" == "true" ] || [ "$(uci get wireguard.mesh_vpn.loadbalancing)" == "1" ]; then
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:101]: Only check for `1` here, that's the canonical uci way of defining booleans.

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 4]
```diff
@@ -65,33 +83,78 @@ if [ "$(uci get wireguard.mesh_vpn.enabled)" == "true" ] || [ "$(uci get wiregua
...
+			# Use /api/v2, get gateway peer details from broker response
+			WGKEX_BROKER="$PROTO://$(uci get wireguard.mesh_vpn.broker)/api/v2/wg/key/exchange"
+			logger -t checkuplink "Contacting wgkex broker $WGKEX_BROKER"
+			WGKEX_DATA=$(gluon-wan wget -q -O- --post-data='{"domain": "'"$SEGMENT"'","public_key": "'"$PUBLICKEY"'"}' "$WGKEX_BROKER")
+			if [ $? -eq 1 ]; then
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:106]: The script will not be able to reach this code within this clause with the latest version of checkuplink (Nov 2023) due to `set -e`. Replace it with `if ! WGKEXDATA=$(...)`.

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 5]
```diff
@@ -65,33 +83,78 @@ if [ "$(uci get wireguard.mesh_vpn.enabled)" == "true" ] || [ "$(uci get wiregua
...
+				logger -t checkuplink "Contacting wgkex broker failed, response: $WGKEX_DATA"
+				exit 1
+			fi
+
+			logger -t checkuplink "Got data from wgkex broker: $WGKEX_DATA"
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:111]: Add a severity to the logger. This looks like `info` level from what I can tell?
https://man7.org/linux/man-pages/man1/logger.1.html#FACILITIES_AND_LEVELS

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 6]
```diff
@@ -65,33 +83,78 @@ if [ "$(uci get wireguard.mesh_vpn.enabled)" == "true" ] || [ "$(uci get wiregua
...
+
+			data=$(lua /lib/gluon/gluon-mesh-wireguard-vxlan/parse-wgkex-response.lua "$WGKEX_DATA")
+			if [ $? -eq 1 ]; then
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:114]: Unreachable code. Replace with `if ! data=$(...)`

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 7]
```diff
@@ -65,33 +83,78 @@ if [ "$(uci get wireguard.mesh_vpn.enabled)" == "true" ] || [ "$(uci get wiregua
...
+				logger -t checkuplink "Parsing wgkex broker data failed"
+				exit 1
+			fi
+
+			logger -t checkuplink "Parsed wgkex broker data"
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:119]: Add a severity to the logger. This looks like debug level even?

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 8]
```diff
@@ -65,33 +83,78 @@ if [ "$(uci get wireguard.mesh_vpn.enabled)" == "true" ] || [ "$(uci get wiregua
...
+
+			PEER_ADDRESS=$(echo "$data" | sed -n 1p)
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:121]: I think a comment where the structure of `data` is defined would be helpful here (or above).

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 9]
```diff
@@ -65,33 +83,78 @@ if [ "$(uci get wireguard.mesh_vpn.enabled)" == "true" ] || [ "$(uci get wiregua
...
+
+		else
+			# Use /api/v2, get gateway peer details from config
+			WGKEX_BROKER="$PROTO://$(uci get wireguard.mesh_vpn.broker)/api/v1/wg/key/exchange"
+			logger -t checkuplink "Contacting wgkex broker $WGKEX_BROKER"
+			gluon-wan wget -q -O- --post-data='{"domain": "'"$SEGMENT"'","public_key": "'"$PUBLICKEY"'"}' "$WGKEX_BROKER"
+			if [ $? -eq 1 ]; then
+				logger -t checkuplink "Contacting wgkex broker failed"
+				exit 1
+			fi
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:137]: Can we deduplicate this code here with the one above?

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 10]
```diff
@@ -65,33 +83,78 @@ if [ "$(uci get wireguard.mesh_vpn.enabled)" == "true" ] || [ "$(uci get wiregua
...
+
+			# Get the number of configured peers and randomly select one
+			NUMBER_OF_PEERS=$(uci -q show wireguard | egrep -ce peer_[0-9]+.endpoint)
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:140]: Some safety checks might be good here. For example checking if `NUMBER_OF_PEERS` is not empty, of checking of `PEER_PUBLICKEY` is not empty, etc.

Assuming someone misconfigures `wireguard.mesh_vpn.loadbalancing` by accident but also has no local wireguard definitions for the peers, we might run into undefined behaviour here.

**Review by grische — CHANGES_REQUESTED** [review]

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 11]
```diff
@@ -28,22 +28,31 @@ clean_port() {
...
 	local gateway
 	gateway="$(clean_port "$peer_endpoint")"
 	# Check if we have a default route for v6 if not fallback to v4
 	if ip -6 route show table 1 | grep -q 'default via'
 	then
 		local ipv6
-		ipv6="$(gluon-wan nslookup "$gateway" | grep 'Address:\? [0-9]' | grep -E -o '([a-f0-9:]+:+)+[a-f0-9]+')"
-		echo "[$ipv6]$(echo "$peer_endpoint" | grep -E -oe ":[0-9]+$")"
+		ipv6="$(gluon-wan nslookup "$gateway" | grep 'Address:\? [0-9]' | grep -oE '([a-f0-9:]+:+)+[a-f0-9]+')"
+		echo "$ipv6"
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:50]: Assuming this regex magic still works, drop the variable `ipv6`.
```suggestion
		gluon-wan nslookup "$gateway" | grep 'Address:\? [0-9]' | grep -oE '([a-f0-9:]+:+)+[a-f0-9]+'
```

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 12]
```diff
@@ -28,22 +28,31 @@ clean_port() {
...
 	else
 		local ipv4
 		ipv4="$(gluon-wan nslookup "$gateway" | grep 'Address:\? [0-9]' | grep -oE "\b([0-9]{1,3}\.){3}[0-9]{1,3}\b")"
-		echo "$ipv4$(echo "$peer_endpoint" | grep -E -oe ":[0-9]+$")"
+		echo "$ipv4"
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:54]: Same as above, just remove the echo.

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink (hunk 1 above)
Comment (grische) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink]: Something wrong happened here during rebase. The function definition of `combine_ip_port` disappeared

**Review by grische — CHANGES_REQUESTED** [review]
@DasSkelett most comments of mine still hold even after the rebase. Especially the unreachable code should be fixed probably.

**Comment by DasSkelett** [comment]
Yes indeed, I only did the rebate so far, which was a necessary step before addressing anything else. I will let you know when I'm ready for another review, I am probably going to mark this PR as "ready for review"/not WIP thrn, please hold off with further reviews until then, it just makes this more convoluted otherwise.

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink (hunk 11 above)
Comment (DasSkelett) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:57]: I find the explicit `echo` quite nice for signalling where data is returned, instead of having to guess where the return value (or whether this function is even supposed to return something).
But I'm okay with removing it if you prefer.

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink (hunk 10 above)
Comment (DasSkelett) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink]: `awk` would return `1` in this case, and all `uci get` requests would fail for the nonexistent element. uci also seems to handle an empty string as "not set" and throws an error, so I think we are relatively safe here.
Especially since this is something that would be set and tested in the firmware before it hits any real node, and very unlikely to be manipulated by a user who doesn't know what he does.

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink (hunk 9 above)
Comment (DasSkelett) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink]: If we want to deduplicate, I'd do it using a function, as not to break up the the if-else blocks, because right now the logic is cleanly separated into two code paths for the different API versions.
Not sure if these few lines, which are mostly error handling, are worth it?

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink (hunk 8 above)
Comment (DasSkelett) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink]: Good idea, added a short description above the call to Lua

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink (hunk 4 above)
Comment (DasSkelett) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink]: Thanks, should be fixed in all cases now

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink (hunk 1 above)
Comment (DasSkelett) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink]: Whoops, readded, and applied your change suggestion

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 13]
```diff
@@ -18,24 +18,42 @@ clean_port() {
...
 
-check_address_family() {
-	local peer_publickey="$1"
-	local peer_endpoint="$2"
-	local gateway="$(clean_port $peer_endpoint)"
+extract_port() {
+    local address_port="$1"
+
+    # Remove brackets for IPv6 addresses
+    local address="$(echo "$address_port" | sed -e 's/\[//;s/\]//')"
```
[the remark this answers is not in the document]
Comment (DasSkelett) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink]: Actually thinking about it and testing it, this line isn't needed at all if I'm not mistaken, the awk command doesn't care about the brackets. Removed this line altogether.

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink (hunk 1 above)
Comment (DasSkelett) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink]: Hm, shellcheck says
```
  ./ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:39:8: error: In dash, [[ ]] is not supported. (ShellCheck.SC3010)
```
Is this true? It does seem to work on a node, tough...

**Comment by DasSkelett** [comment]
@grische please do one more pass, if it looks good to you I'll reopen this PR upstream

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 14]
```diff
@@ -75,25 +86,24 @@ if [ "$(uci get wireguard.mesh_vpn.enabled)" = "true" ] || [ "$(uci get wireguar
...
 		logger -t checkuplink "Reconnecting ..."
 		NTP_SERVERS=$(uci get system.ntp.server)
-		# shellcheck disable=SC3060  # busybox sh supports string replacement
-		NTP_SERVERS="${NTP_SERVERS// / -p }"  # each separate NTP server needs to be behind a "-p"
-		# shellcheck disable=SC2086  # we need to expand the list of NTP_SERVERS here
-		if ! gluon-wan /usr/sbin/ntpd -n -N -S /usr/sbin/ntpd-hotplug -p ${NTP_SERVERS} -q
+		NTP_SERVERS_ADDRS=""
+		for NTP_SERVER in $NTP_SERVERS; do
+			ipv6="$(gluon-wan nslookup "$NTP_SERVER" | grep 'Address:\? [0-9]' | grep -E -o '([a-f0-9:]+:+)+[a-f0-9]+')"
+			ipv4="$(gluon-wan nslookup "$NTP_SERVER" | grep 'Address:\? [0-9]' | grep -oE "\b([0-9]{1,3}\.){3}[0-9]{1,3}\b")"
```
Comment (T0biii) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:92]: maybe call gluon-wan once?

```
addresses=$(gluon-wan nslookup "$NTP_SERVER" | grep 'Address:\? [0-9]')
ipv6=$(echo "$addresses" | grep -E -o '([a-f0-9:]+:+)+[a-f0-9]+')
ipv4=$(echo "$addresses" | grep -oE "\b([0-9]{1,3}\.){3}[0-9]{1,3}\b")
```

**Review by T0biii — COMMENTED** [review]
maybe got one idea for a change

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink (hunk 14 above)
Comment (grische) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:92]: as gluon-wan also was the source of the problems in the past, it seems best to have it separate in case we need to debug this further (i.e. run it with `sh -x checkuplink`).

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 15]
```diff
@@ -27,22 +27,33 @@ clean_port() {
...
 		local ipv6
-		ipv6="$(gluon-wan nslookup "$gateway" | grep 'Address:\? [0-9]' | grep -E -o '([a-f0-9:]+:+)+[a-f0-9]+')"
-		echo "[$ipv6]$(echo "$peer_endpoint" | grep -E -oe ":[0-9]+$")"
+		ipv6="$(gluon-wan nslookup "$gateway" | grep 'Address:\? [0-9]' | grep -oE '([a-f0-9:]+:+)+[a-f0-9]+')"
+		echo "$ipv6"
 	else
 		local ipv4
-		ipv4="$(gluon-wan nslookup "$gateway" | grep 'Address:\? [0-9]' | grep -oE "\b([0-9]{1,3}\.){3}[0-9]{1,3}\b")"
-		echo "$ipv4$(echo "$peer_endpoint" | grep -E -oe ":[0-9]+$")"
+		ipv4="$(gluon-wan nslookup "$gateway" | grep 'Address:\? [0-9]' | grep -oE '\b([0-9]{1,3}\.){3}[0-9]{1,3}\b')"
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:54]: Is this change necessary? While I am not sure if `'\b'` and `"\b"` are strictly different, `'\\b'` and `"\\b"` for example are,

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 16]
```diff
@@ -114,7 +124,64 @@ if [ "$(uci get wireguard.mesh_vpn.enabled)" = "true" ] || [ "$(uci get wireguar
...
-		gluon-wan wget -q  -O- --post-data='{"domain": "'"$SEGMENT"'","public_key": "'"$PUBLICKEY"'"}' "$PROTO://$(uci get wireguard.mesh_vpn.broker)"
+
+		# Remove API path suffix if still present in config
+		WGKEX_BROKER_BASE_PATH="$(uci get wireguard.mesh_vpn.broker | sed 's|/api/v1/wg/key/exchange||')"
+
+		if [ "$(uci get wireguard.mesh_vpn.loadbalancing)" = "1" ]; then
+			# Use /api/v2, get gateway peer details from broker response
+			WGKEX_BROKER="$PROTO://$WGKEX_BROKER_BASE_PATH/api/v2/wg/key/exchange"
+			logger -t checkuplink "Contacting wgkex broker $WGKEX_BROKER"
+			if ! WGKEX_DATA=$(LD_PRELOAD=libpacketmark.so LIBPACKETMARK_MARK=1 gluon-wan wget -q -O- --post-data='{"domain": "'"$SEGMENT"'","public_key": "'"$PUBLICKEY"'"}' "$WGKEX_BROKER"); then
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:135]: It would be nice if we could move this LD_PRELOAD into a function, as we are using this in several places now.

**Review by grische — CHANGES_REQUESTED** [review]
The only real issue I can see is this `[[ ${ip} == *":"* ]];`, otherwise a few things we can also improve later.

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink (hunk 9 above)
Comment (grische) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink]: We can also do this later if you want.

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink (hunk 14 above)
Comment (DasSkelett) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:92]: This is from https://github.com/freifunk-gluon/community-packages/pull/92 (I don't know why GitHub stil shows this as change in this PR, the base branch already contains the commit).
Should be fixed separately, not here.

Comment (grische) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink:92]: @DasSkelett is it possible that your last rebase is not based on the `main` from this repo?

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink (hunk 16 above)
Comment (grische) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink]: I added a commit that does this

File: ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink [hunk 17]
```diff
@@ -27,22 +27,33 @@ clean_port() {
...
-	gateway="$(clean_port "$peer_endpoint")"
+extract_port() {
+    echo "$1" | awk -F: '{print $NF}'
+}
+
+combine_ip_port() {
+    local ip="$1"
+    local port="$2"
+
+    if [[ ${ip} == *":"* ]]; then
```
[the remark this answers is not in the document]
Comment (grische) [reply ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink]: I fixed this

**Comment by grische** [comment]
The latest commit also fixes the NTP issues that were caused occasionally be some IPv6 regex mismatch. See https://github.com/freifunk-gluon/community-packages/pull/96 for details

**Comment by grische** [comment]
Fixed the merge conflicts

File: ffmuc-mesh-vpn-wireguard-vxlan/README.md [hunk 18]
```diff
@@ -1,53 +1,58 @@
...
-					link_address = 'fe80::f000:22ff:fe12:02',
-				},
-				{
-					publickey ='xakSGG39D1v90j3Z9eVWzojh6nDbnsVUc/RByVdcKB0=',
-					endpoint ='ffkwsn03.freifunk-koenigswinter.de:30020',
-					link_address = 'fe80::f000:22ff:fe12:07',
-				},
-
+		broker = 'broker.ffmuc.net', -- base path of broker, will be combined with API path
+		loadbalancing = '1' -- controls whether to use the loadbalancing/gateway assignment feature of the broker
```
Comment (grische) [inline ffmuc-mesh-vpn-wireguard-vxlan/README.md:34]: I think there's a comma missing here

**Comment by grische** [comment]
I considerably rewrote the whole function so the integration actually works as intended:
https://github.com/freifunkMUC/community-packages/compare/4bf62bcbe723aa3c69e6cbccd38d0b8e213e4d7b..b82299dd660158a4240053d1a88bc94f15711e70

This needs a proper re-test.
